flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/shuffle/KafkaShuffleITCase.java

```diff
@@ -56,6 +56,7 @@
 import static org.apache.flink.streaming.connectors.kafka.shuffle.FlinkKafkaShuffle.PARTITION_NUMBER;
 import static org.apache.flink.streaming.connectors.kafka.shuffle.FlinkKafkaShuffle.PRODUCER_PARALLELISM;
 import static org.apache.flink.test.util.TestUtils.tryExecute;
+import static org.junit.Assert.assertEquals;
```

Comment:
Could you split unrelated changes (`assertEquals`, or some changes in comments) from this bug fix commit?

Reply:
Changes in comments are related to the fix. There is a slight change in behaviour for `ProcessingTime` that we emit the `MAX_WATERMARK`. Previously it was filtered by the `StreamSourceContext`.

I moved emitting the `MAX_WATERMARK` out from the `StreamSource` (operator) to the `StreamSourceTask` (task). I did it because if we restore a fully finished task the StreamSourceContext is not initialised when calling `advanceToEndOfTime`. We skip life cycle methods of an operator in such a case.

I think this change should be fine though. It is also how we do it in new sources.